### PR TITLE
Use `--code-coverage=lcov-%p.info` flag to collect code coverage

### PR DESCRIPTION
**Pull request by giordano**
Fix https://github.com/JuliaLang/julia/issues/46048

### CODE REVIEW

**Comment by vtjnash** [comment]
You might also need to fix the uploader script to know where to look for this file. And resign the pipelines.

**Comment by DilumAluthge** [comment]
You'll also need to temporarily modify lines 6 and 7 of https://github.com/JuliaCI/julia-buildkite/blob/main/pipelines/main/coverage/coverage.yml#L7 so that you can test the changes in this PR.

**Comment by DilumAluthge** [comment]
> You might also need to fix the uploader script to know where to look for this file.

The uploader script simply calls `Coverage.process_folder("base")` and `Coverage.process_folder("stdlib")`. Will `Coverage.process_folder` automatically find the new files?

**Comment by DilumAluthge** [comment]
From looking at https://github.com/JuliaCI/CoverageTools.jl/blob/9d8fba371f4c65a7e9ca3bdaae753a52bcb22218/src/CoverageTools.jl#L274, I think what we really want is for the coverage files to be named `sourcefile.jl.$PID.cov`, right? Is there a way that we can accomplish this using the `--code-coverage` command line flag?

**Comment by vtjnash** [comment]
That emits lcov files, so you use `LCOV.readfolder()` instead on some readable tmpdir

**Comment by DilumAluthge** [comment]
@vtjnash It looks like the coverage info from different files isn't being merged correctly?

**Comment by DilumAluthge** [comment]
E.g. if you look at the log, you will see output like:

```
[ Info: /cache/build/default-amdci4-4/julialang/julia-buildkite/usr/share/julia/stdlib/v1.9/Test/src/logging.jl 74 %            41             55
[ Info: /cache/build/default-amdci4-4/julialang/julia-buildkite/usr/share/julia/stdlib/v1.9/Test/src/logging.jl 87 %            48             55
[ Info: /cache/build/default-amdci4-4/julialang/julia-buildkite/usr/share/julia/stdlib/v1.9/Test/src/logging.jl 74 %            41             55
[ Info: /cache/build/default-amdci4-4/julialang/julia-buildkite/usr/share/julia/stdlib/v1.9/Test/src/logging.jl 71 %            48             67
[ Info: /cache/build/default-amdci4-4/julialang/julia-buildkite/usr/share/julia/stdlib/v1.9/Test/src/logging.jl 92 %            62             67
```

But shouldn't those all be merged into a single `FileCoverage` element?

(This is when iterating over each element of `fcs` and printing. `fcs` is the `Vector{FileCoverage}` that you get when you do `fcs = Coverage.LCOV.readfolder(".")`.

**Comment by DilumAluthge** [comment]
Also, https://github.com/JuliaLang/julia/blob/master/stdlib/Test/src/logging.jl has 336 lines, not 55 or 67.

**Comment by vtjnash** [comment]
You might want to read the docs, it is right at the top how to use this: https://github.com/JuliaCI/Coverage.jl#code-coverage

that sounds like covered/relevant lines, before processing

File: pipelines/scheduled/coverage/coverage.yml [hunk 1]
```diff
@@ -5,7 +5,8 @@ steps:
         key:   "coverage"
         # Comment out the following `if:` statement for testing.
         # Please remember to uncomment it before merging the PR.
-        if: pipeline.slug == "julia-master-scheduled"
+        # if: build.source == "schedule" # comment out this line for testing; please uncomment it before merging your PR
+        if: (build.source == "schedule") || (pipeline.slug == "julia-buildkite") || (pipeline.slug == "julia-buildkite-scheduled") # uncomment this line for testing; please comment it out before merging your PR
```
Comment (DilumAluthge) [inline pipelines/scheduled/coverage/coverage.yml:9]: These lines are outdated now that we have moved to a separate scheduled pipeline.

Take a look at how we do it on master: https://github.com/JuliaCI/julia-buildkite/blob/main/pipelines/scheduled/coverage/coverage.yml

Roughly speaking, you want to delete the two `if` lines (i.e. delete lines 8 and 9) and replace them with the following three lines:

```yml
        # Comment out the following `if:` statement for testing.
        # Please remember to uncomment it before merging the PR.
        if: pipeline.slug == "julia-master-scheduled"
```

**Comment by DilumAluthge** [comment]
On the latest `coverage` job on this PR, I see the following:

```
[ Info: Total: 94% (36228/38519)
```

@staticfloat @vtjnash This is suspiciously high, right? IIRC, we were previously 87-88% pretty consistently.

**Comment by inkydragon** [comment]
> This is suspiciously high, right?

Maybe this is because we only uploaded coverage data for stdlib.

https://codecov.io/gh/JuliaLang/julia/tree/master/cache/build/default-amdci4-5/julialang/julia-buildkite-scheduled/usr/share/julia

**Comment by staticfloat** [comment]
Do we know why so many tests are failing here?

**Comment by DilumAluthge** [comment]
As I understand it, these test failures are expected when code coverage emission is enabled.

**Comment by DilumAluthge** [comment]
Note that we allow the code coverage job to be "green" on Buildkite even if some of the tests fail.

**Comment by staticfloat** [comment]
Here's the uploaded report: https://app.codecov.io/gh/JuliaLang/julia/commit/57ffed6bf9edae35e6d27a8704622c83143ae214

It looks good to me, I'm gonna merge as soon as `Check` passes.

**Comment by giordano** [comment]
I'm very proud of the hard work I put into this PR

**Comment by vtjnash** [comment]
This is a great step! But there seem to be a few issues still, looking at https://app.codecov.io/gh/JuliaLang/julia

- [ ] paths are missing `base/` prefix, so the coverage tools can't find them in the repo
- [ ] there is a test file accidentally present https://codecov.io/gh/JuliaLang/julia/src/master/cache/build/default-amdci5-7/julialang/julia-master-scheduled/usr/share/julia/test/project/deps/Foo1/src/Foo.jl
- [ ] the stdlib files have their full path instead of being `stdlib/` prefixed https://codecov.io/gh/JuliaLang/julia/tree/master/cache/build/default-amdci5-7/julialang/julia-master-scheduled/usr/share/julia/stdlib/v1.9
- [ ] Pkg (and other similar files that aren't part of Julia) are getting included in the stats
- [ ] coverage changed from ~89% to ~93%, which seems unlikely and might indicate the `amend_coverage_from_src!` step is being skipped, though it is hard to compare what changed currently do to previous issues

**Comment by DilumAluthge** [comment]
@vtjnash I moved your comment to an issue so that we can track each of the bullet points: #188

**Pull request merged**